src/NodaTime/Duration.cs

```diff
@@ -172,6 +172,41 @@ private Duration(long units, string paramName, long minValue, long maxValue, lon
             }
         }
 
+        // Note: unitsPerDay is always actually an integer, but we always want to perform floating point arithmetic
+        // with it.
+        private Duration(double units, string paramName, double unitsPerDay, long nanosPerUnit)
+        {
+            // Note: we don't check the original double values other than for NaN and infinity.
+            // Instead, we do the relevant conversion and then check that we're in bounds.
+            // That way we won't get invalid values due to weird rounding issues.
+            if (double.IsNaN(units) || double.IsInfinity(units))
+            {
+                throw new ArgumentOutOfRangeException(paramName, "Value must be a finite, valid number.");
+            }
+            unchecked
+            {
+                days = (int)(units / unitsPerDay);
+                double unitOfDay = units - (unitsPerDay * days);
+                if (unitOfDay < 0)
```

Comment:
Just checking: is this reachable? (My guess is that it is, but only when the space between consecutive `double` values exceeds `unitsPerDay`, but I'm still struggling to see how we get a negative value here, even then.)

Reply:
Hmm... will check when I get a chance. (Probably not for a while.) We've got the same thing in the earlier integer version too.